File: src/parser.zig
const std = @import("std");
const lexer = @import("lexer.zig");
const ast = @import("ast.zig");
const types = @import("types.zig");

const Token = lexer.Token;
const TokenType = lexer.TokenType;

pub const ParseError = error{
    UnexpectedToken,
    ExpectedExpression,
    ExpectedStatement,
    ExpectedIdentifier,
    ExpectedType,
    OutOfMemory,
};

pub const Parser = struct {
    const Self = @This();

    allocator: std.mem.Allocator,
    tokens: []Token,
    current: usize,

    pub fn init(allocator: std.mem.Allocator, tokens: []Token) Self {
        return Self{
            .allocator = allocator,
            .tokens = tokens,
            .current = 0,
        };
    }

    pub fn deinit(self: *Self) void {
        _ = self;
    }

    pub fn parseProgram(self: *Self) !*ast.Node {
        var statements = std.ArrayList(*ast.Node){};
        defer statements.deinit(self.allocator);

        while (!self.isAtEnd()) {
            if (self.check(.eof)) break;
            const stmt = try self.declaration();
            try statements.append(self.allocator,stmt);
        }

        return ast.createProgram(self.allocator, try statements.toOwnedSlice(self.allocator));
    }

    fn declaration(self: *Self) ParseError!*ast.Node {
        if (self.match(.fn_)) return self.functionDeclaration();
        if (self.match(.const_)) return self.variableDeclaration(false);
        if (self.match(.let_)) return self.variableDeclaration(true);
        return self.statement();
    }

    fn functionDeclaration(self: *Self) ParseError!*ast.Node {
        const name_token = try self.consume(.identifier, "Expected function name");
        const name = name_token.lexeme;

        _ = try self.consume(.left_paren, "Expected '(' after function name");

        var parameters = std.ArrayList(ast.Parameter){};
        defer parameters.deinit(self.allocator);

        if (!self.check(.right_paren)) {
            while (true) {
                const param_name_token = try self.consume(.identifier, "Expected parameter name");
                _ = try self.consume(.colon, "Expected ':' after parameter name");
                const param_type = try self.parseType();

                try parameters.append(self.allocator,ast.Parameter{
                    .name = param_name_token.lexeme,
                    .param_type = param_type,
                });

                if (!self.match(.comma)) break;
            }
        }

        _ = try self.consume(.right_paren, "Expected ')' after parameters");
        _ = try self.consume(.arrow, "Expected '->' after parameters");
        const return_type = try self.parseType();

        _ = try self.consume(.left_brace, "Expected '{' before function body");
        const body = try self.blockStatement();

        return ast.createFunctionDeclaration(
            self.allocator,
            getPosition(name_token),
            name,
            try parameters.toOwnedSlice(self.allocator),
            return_type,
            body,
        );
    }

    fn variableDeclaration(self: *Self, is_mutable: bool) ParseError!*ast.Node {
        const name_token = try self.consume(.identifier, "Expected variable name");

        var var_type: ?types.Type = null;
        if (self.match(.colon)) {
            var_type = try self.parseType();
        }

        var initializer: ?*ast.Node = null;
        if (self.match(.equal)) {
            initializer = try self.expression();
        }

        _ = try self.consume(.semicolon, "Expected ';' after variable declaration");

        const node = try ast.createNode(self.allocator, .variable_declaration, getPosition(name_token));
        node.data = .{ .variable_declaration = ast.VariableDeclaration{
            .name = name_token.lexeme,
            .var_type = var_type,
            .is_mutable = is_mutable,
            .initializer = initializer,
        } };
        return node;
    }

    fn statement(self: *Self) ParseError!*ast.Node {
        if (self.match(.if_)) return self.ifStatement();
        if (self.match(.while_)) return self.whileStatement();
        if (self.match(.return_)) return self.returnStatement();
        if (self.match(.left_brace)) return self.blockStatement();
        return self.expressionStatement();
    }

    fn ifStatement(self: *Self) ParseError!*ast.Node {
        _ = try self.consume(.left_paren, "Expected '(' after 'if'");
        const condition = try self.expression();
        _ = try self.consume(.right_paren, "Expected ')' after if condition");

        const then_branch = try self.statement();
        var else_branch: ?*ast.Node = null;
        if (self.match(.else_)) {
            else_branch = try self.statement();
        }

        const node = try ast.createNode(self.allocator, .if_statement, getPosition(self.previous()));
        node.data = .{ .if_statement = ast.IfStatement{
            .condition = condition,
            .then_branch = then_branch,
            .else_branch = else_branch,
        } };
        return node;
    }

    fn whileStatement(self: *Self) ParseError!*ast.Node {
        _ = try self.consume(.left_paren, "Expected '(' after 'while'");
        const condition = try self.expression();
        _ = try self.consume(.right_paren, "Expected ')' after while condition");

        const body = try self.statement();

        const node = try ast.createNode(self.allocator, .while_statement, getPosition(self.previous()));
        node.data = .{ .while_statement = ast.WhileStatement{
            .condition = condition,
            .body = body,
        } };
        return node;
    }

    fn returnStatement(self: *Self) ParseError!*ast.Node {
        const keyword = self.previous();
        var value: ?*ast.Node = null;

        if (!self.check(.semicolon)) {
            value = try self.expression();
        }

        _ = try self.consume(.semicolon, "Expected ';' after return value");

        const node = try ast.createNode(self.allocator, .return_statement, getPosition(keyword));
        node.data = .{ .return_statement = ast.ReturnStatement{ .value = value } };
        return node;
    }

    fn blockStatement(self: *Self) ParseError!*ast.Node {
        var statements = std.ArrayList(*ast.Node){};
        defer statements.deinit(self.allocator);

        while (!self.check(.right_brace) and !self.isAtEnd()) {
            try statements.append(self.allocator,try self.declaration());
        }

        _ = try self.consume(.right_brace, "Expected '}' after block");

        const node = try ast.createNode(self.allocator, .block_statement, getPosition(self.previous()));
        node.data = .{ .block_statement = ast.BlockStatement{
            .statements = try statements.toOwnedSlice(self.allocator),
        } };
        return node;
    }

    fn expressionStatement(self: *Self) ParseError!*ast.Node {
        const expr = try self.expression();
        _ = try self.consume(.semicolon, "Expected ';' after expression");

        const node = try ast.createNode(self.allocator, .expression_statement, getPosition(self.previous()));
        node.data = .{ .expression_statement = ast.ExpressionStatement{ .expression = expr } };
        return node;
    }

    fn expression(self: *Self) ParseError!*ast.Node {
        return self.assignment();
    }

    fn assignment(self: *Self) ParseError!*ast.Node {
        const expr = try self.logicalOr();

        if (self.match(.equal)) {
            const value = try self.assignment();
            if (expr.node_type == .identifier) {
                const node = try ast.createNode(self.allocator, .assignment_expression, getPosition(self.previous()));
                node.data = .{ .assignment_expression = ast.AssignmentExpression{
                    .target = expr,
                    .value = value,
                } };
                return node;
            }
            return ParseError.ExpectedExpression;
        }

        return expr;
    }

    fn logicalOr(self: *Self) ParseError!*ast.Node {
        var expr = try self.logicalAnd();

        while (self.match(.equal_equal)) { // Using == as OR for now, add proper || later
            const operator: ast.BinaryOperator = .logical_or;
            const right = try self.logicalAnd();
            const node = try ast.createNode(self.allocator, .binary_expression, getPosition(self.previous()));
            node.data = .{ .binary_expression = ast.BinaryExpression{
                .left = expr,
                .operator = operator,
                .right = right,
            } };
            expr = node;
        }

        return expr;
    }

    fn logicalAnd(self: *Self) ParseError!*ast.Node {
        const expr = try self.equality();

        // Add && support later
        return expr;
    }

    fn equality(self: *Self) ParseError!*ast.Node {
        var expr = try self.comparison();

        while (self.match(.bang_equal) or self.match(.equal_equal)) {
            const operator: ast.BinaryOperator = if (self.previous().type == .bang_equal) .not_equal else .equal;
            const right = try self.comparison();
            const node = try ast.createNode(self.allocator, .binary_expression, getPosition(self.previous()));
            node.data = .{ .binary_expression = ast.BinaryExpression{
                .left = expr,
                .operator = operator,
                .right = right,
            } };
            expr = node;
        }

        return expr;
    }

    fn comparison(self: *Self) ParseError!*ast.Node {
        var expr = try self.term();

        while (self.match(.greater) or self.match(.greater_equal) or self.match(.less) or self.match(.less_equal)) {
            const operator: ast.BinaryOperator = switch (self.previous().type) {
                .greater => .greater_than,
                .greater_equal => .greater_equal,
                .less => .less_than,
                .less_equal => .less_equal,
                else => unreachable,
            };
            const right = try self.term();
            const node = try ast.createNode(self.allocator, .binary_expression, getPosition(self.previous()));
            node.data = .{ .binary_expression = ast.BinaryExpression{
                .left = expr,
                .operator = operator,
                .right = right,
            } };
            expr = node;
        }

        return expr;
    }

    fn term(self: *Self) ParseError!*ast.Node {
        var expr = try self.factor();

        while (self.match(.minus) or self.match(.plus)) {
            const operator: ast.BinaryOperator = if (self.previous().type == .minus) .subtract else .add;
            const right = try self.factor();
            const node = try ast.createNode(self.allocator, .binary_expression, getPosition(self.previous()));
            node.data = .{ .binary_expression = ast.BinaryExpression{
                .left = expr,
                .operator = operator,
                .right = right,
            } };
            expr = node;
        }

        return expr;
    }

    fn factor(self: *Self) ParseError!*ast.Node {
        var expr = try self.unary();

        while (self.match(.slash) or self.match(.star)) {
            const operator: ast.BinaryOperator = if (self.previous().type == .slash) .divide else .multiply;
            const right = try self.unary();
            const node = try ast.createNode(self.allocator, .binary_expression, getPosition(self.previous()));
            node.data = .{ .binary_expression = ast.BinaryExpression{
                .left = expr,
                .operator = operator,
                .right = right,
            } };
            expr = node;
        }

        return expr;
    }

    fn unary(self: *Self) ParseError!*ast.Node {
        if (self.match(.bang) or self.match(.minus)) {
            const operator: ast.UnaryOperator = if (self.previous().type == .bang) .logical_not else .minus;
            const right = try self.unary();
            const node = try ast.createNode(self.allocator, .unary_expression, getPosition(self.previous()));
            node.data = .{ .unary_expression = ast.UnaryExpression{
                .operator = operator,
                .operand = right,
            } };
            return node;
        }

        return self.call();
    }

    fn call(self: *Self) ParseError!*ast.Node {
        var expr = try self.primary();

        while (true) {
            if (self.match(.left_paren)) {
                expr = try self.finishCall(expr);
            } else {
                break;
            }
        }

        return expr;
    }

    fn finishCall(self: *Self, callee: *ast.Node) ParseError!*ast.Node {
        var arguments = std.ArrayList(*ast.Node){};
        defer arguments.deinit(self.allocator);

        if (!self.check(.right_paren)) {
            while (true) {
                try arguments.append(self.allocator,try self.expression());
                if (!self.match(.comma)) break;
            }
        }

        _ = try self.consume(.right_paren, "Expected ')' after arguments");

        const node = try ast.createNode(self.allocator, .call_expression, getPosition(self.previous()));
        node.data = .{ .call_expression = ast.CallExpression{
            .function = callee,
            .arguments = try arguments.toOwnedSlice(self.allocator),
        } };
        return node;
    }

    fn primary(self: *Self) ParseError!*ast.Node {
        if (self.match(.false_)) {
            const node = try ast.createNode(self.allocator, .boolean_literal, getPosition(self.previous()));
            node.data = .{ .boolean_literal = ast.BooleanLiteral{ .value = false } };
            return node;
        }

        if (self.match(.true_)) {
            const node = try ast.createNode(self.allocator, .boolean_literal, getPosition(self.previous()));
            node.data = .{ .boolean_literal = ast.BooleanLiteral{ .value = true } };
            return node;
        }

        if (self.match(.number)) {
            const lexeme = self.previous().lexeme;
            const value = std.fmt.parseFloat(f64, lexeme) catch return ParseError.ExpectedExpression;
            const is_integer = std.mem.indexOf(u8, lexeme, ".") == null;
            return ast.createNumberLiteral(self.allocator, getPosition(self.previous()), value, is_integer);
        }

        if (self.match(.string_literal)) {
            const node = try ast.createNode(self.allocator, .string_literal, getPosition(self.previous()));
            node.data = .{ .string_literal = ast.StringLiteral{ .value = self.previous().lexeme } };
            return node;
        }

        if (self.match(.identifier)) {
            return ast.createIdentifier(self.allocator, getPosition(self.previous()), self.previous().lexeme);
        }

        if (self.match(.left_paren)) {
            const expr = try self.expression();
            _ = try self.consume(.right_paren, "Expected ')' after expression");
            return expr;
        }

        return ParseError.ExpectedExpression;
    }

    fn parseType(self: *Self) ParseError!types.Type {
        if (self.match(.void)) return .void;
        if (self.match(.i32)) return .i32;
        if (self.match(.f64)) return .f64;
        if (self.match(.bool)) return .bool;
        if (self.match(.string)) return .string;
        return ParseError.ExpectedType;
    }

    // Utility methods
    fn match(self: *Self, token_type: TokenType) bool {
        if (self.check(token_type)) {
            _ = self.advance();
            return true;
        }
        return false;
    }

    fn check(self: *Self, token_type: TokenType) bool {
        if (self.isAtEnd()) return false;
        return self.peek().type == token_type;
    }

    fn advance(self: *Self) Token {
        if (!self.isAtEnd()) self.current += 1;
        return self.previous();
    }

    fn isAtEnd(self: *Self) bool {
        return self.peek().type == .eof;
    }

    fn peek(self: *Self) Token {
        return self.tokens[self.current];
    }

    fn previous(self: *Self) Token {
        return self.tokens[self.current - 1];
    }

    fn consume(self: *Self, token_type: TokenType, message: []const u8) ParseError!Token {
        if (self.check(token_type)) return self.advance();

        std.debug.print("Parse error: {s} at token '{s}'\n", .{ message, self.peek().lexeme });
        return ParseError.UnexpectedToken;
    }
};

// Token helper method
fn tokenGetPosition(token: Token) ast.Position {
    return ast.Position{ .line = token.line, .column = token.column };
}

// Add method to Token type (this is a workaround since we can't extend the Token struct directly)
const TokenExtensions = struct {
    fn getPosition(token: Token) ast.Position {
        return ast.Position{ .line = token.line, .column = token.column };
    }
};

// Make the extension available
pub fn getPosition(token: Token) ast.Position {
    return TokenExtensions.getPosition(token);
}

File: src/parser/parser.cpp
#include "parser.h"
#include <iostream>
#include <stdexcept>

namespace zen {

Parser::Parser(std::vector<std::unique_ptr<Token>> tokens) 
    : tokens(std::move(tokens)), current(0) {}

std::unique_ptr<ASTNode> Parser::parse() {
    try {
        return program();
    } catch (const std::exception& e) {
        std::cerr << "Parse error: " << e.what() << std::endl;
        return nullptr;
    }
}

Token* Parser::peek(size_t offset) const {
    size_t index = current + offset;
    if (index >= tokens.size()) {
        return tokens.back().get(); // EOF token
    }
    return tokens[index].get();
}

Token* Parser::advance() {
    if (!isAtEnd()) current++;
    return tokens[current - 1].get();
}

bool Parser::isAtEnd() const {
    return peek()->type == TokenType::END_OF_FILE;
}

bool Parser::check(TokenType type) const {
    if (isAtEnd()) return false;
    return peek()->type == type;
}

bool Parser::match(std::initializer_list<TokenType> types) {
    for (TokenType type : types) {
        if (check(type)) {
            advance();
            return true;
        }
    }
    return false;
}

Token* Parser::consume(TokenType type, const std::string& message) {
    if (check(type)) return advance();
    
    error(message);
    throw std::runtime_error(message);
}

void Parser::synchronize() {
    advance();
    
    while (!isAtEnd()) {
        if (tokens[current - 1]->type == TokenType::SEMICOLON) return;
        
        switch (peek()->type) {
            case TokenType::FUNC:
            case TokenType::LET:
            case TokenType::CONST:
            case TokenType::IF:
            case TokenType::WHILE:
            case TokenType::FOR:
            case TokenType::RETURN:
                return;
            default:
                break;
        }
        
        advance();
    }
}

void Parser::error(const std::string& message) {
    Token* token = peek();
    std::cerr << "Error at line " << token->line << ", column " << token->column 
              << ": " << message << std::endl;
}

std::unique_ptr<ASTNode> Parser::program() {
    auto program = std::make_unique<ProgramNode>();
    
    while (!isAtEnd()) {
        if (match({TokenType::NEWLINE})) {
            continue; // Skip newlines at top level
        }
        
        auto decl = declaration();
        if (decl) {
            program->declarations.push_back(std::move(decl));
        }
    }
    
    return std::move(program);
}

std::unique_ptr<ASTNode> Parser::declaration() {
    try {
        if (match({TokenType::FUNC})) return functionDeclaration();
        if (match({TokenType::LET, TokenType::CONST})) return variableDeclaration();
        
        return statement();
    } catch (const std::exception& e) {
        synchronize();
        return nullptr;
    }
}

std::unique_ptr<ASTNode> Parser::functionDeclaration() {
    Token* name = consume(TokenType::IDENTIFIER, "Expected function name");
    
    consume(TokenType::LEFT_PAREN, "Expected '(' after function name");
    std::vector<Parameter> parameters = parseParameters();
    consume(TokenType::RIGHT_PAREN, "Expected ')' after parameters");
    
    std::string return_type = "void";
    if (match({TokenType::ARROW})) {
        return_type = parseType();
    }
    
    consume(TokenType::LEFT_BRACE, "Expected '{' before function body");
    auto body = blockStatement();
    
    auto func = std::make_unique<FunctionDeclarationNode>(
        name->lexeme, return_type, std::move(body), false, name->line, name->column
    );
    func->parameters = std::move(parameters);
    
    return std::move(func);
}

std::unique_ptr<ASTNode> Parser::variableDeclaration() {
    bool is_constant = tokens[current - 1]->type == TokenType::CONST;
    
    Token* name = consume(TokenType::IDENTIFIER, "Expected variable name");
    
    consume(TokenType::COLON, "Expected ':' after variable name");
    std::string type_name = parseType();
    
    std::unique_ptr<ASTNode> initializer = nullptr;
    if (match({TokenType::ASSIGN})) {
        initializer = expression();
    }
    
    if (!is_constant && !initializer) {
        error("Variable declaration requires initializer");
    }
    
    consume(TokenType::SEMICOLON, "Expected ';' after variable declaration");
    
    return std::make_unique<VariableDeclarationNode>(
        name->lexeme, type_name, std::move(initializer), is_constant,
        false, name->line, name->column
    );
}

std::unique_ptr<ASTNode> Parser::statement() {
    if (match({TokenType::IF})) return ifStatement();
    if (match({TokenType::WHILE})) return whileStatement();
    if (match({TokenType::RETURN})) return returnStatement();
    if (match({TokenType::LEFT_BRACE})) return blockStatement();
    
    return expressionStatement();
}

std::unique_ptr<ASTNode> Parser::blockStatement() {
    auto block = std::make_unique<BlockStatementNode>();
    
    while (!check(TokenType::RIGHT_BRACE) && !isAtEnd()) {
        if (match({TokenType::NEWLINE})) {
            continue; // Skip newlines in blocks
        }
        
        auto stmt = declaration();
        if (stmt) {
            block->statements.push_back(std::move(stmt));
        }
    }
    
    consume(TokenType::RIGHT_BRACE, "Expected '}' after block");
    return std::move(block);
}

std::unique_ptr<ASTNode> Parser::expressionStatement() {
    auto expr = expression();
    consume(TokenType::SEMICOLON, "Expected ';' after expression");
    return std::make_unique<ExpressionStmtNode>(std::move(expr));
}

std::unique_ptr<ASTNode> Parser::ifStatement() {
    consume(TokenType::LEFT_PAREN, "Expected '(' after 'if'");
    auto condition = expression();
    consume(TokenType::RIGHT_PAREN, "Expected ')' after if condition");
    
    auto then_branch = statement();
    std::unique_ptr<ASTNode> else_branch = nullptr;
    
    if (match({TokenType::ELSE})) {
        else_branch = statement();
    }
    
    return std::make_unique<IfStatementNode>(
        std::move(condition), std::move(then_branch), std::move(else_branch)
    );
}

std::unique_ptr<ASTNode> Parser::whileStatement() {
    consume(TokenType::LEFT_PAREN, "Expected '(' after 'while'");
    auto condition = expression();
    consume(TokenType::RIGHT_PAREN, "Expected ')' after while condition");
    
    auto body = statement();
    
    // For now, create a simple while node (would need WhileStatementNode in AST)
    return std::make_unique<BlockStatementNode>();
}

std::unique_ptr<ASTNode> Parser::returnStatement() {
    Token* keyword = tokens[current - 1].get();
    std::unique_ptr<ASTNode> value = nullptr;
    
    if (!check(TokenType::SEMICOLON)) {
        value = expression();
    }
    
    consume(TokenType::SEMICOLON, "Expected ';' after return value");
    return std::make_unique<ReturnStatementNode>(std::move(value), keyword->line, keyword->column);
}

std::unique_ptr<ASTNode> Parser::expression() {
    return assignment();
}

std::unique_ptr<ASTNode> Parser::assignment() {
    auto expr = logicalOr();
    
    if (match({TokenType::ASSIGN, TokenType::PLUS_ASSIGN, TokenType::MINUS_ASSIGN,
               TokenType::MULTIPLY_ASSIGN, TokenType::DIVIDE_ASSIGN})) {
        Token* operator_token = tokens[current - 1].get();
        auto value = assignment();
        
        return std::make_unique<AssignmentNode>(
            std::move(expr), std::move(value), operator_token->type,
            operator_token->line, operator_token->column
        );
    }
    
    return expr;
}

std::unique_ptr<ASTNode> Parser::logicalOr() {
    auto expr = logicalAnd();
    
    while (match({TokenType::OR})) {
        Token* operator_token = tokens[current - 1].get();
        auto right = logicalAnd();
        expr = std::make_unique<BinaryOpNode>(
            std::move(expr), std::move(right), operator_token->type,
            operator_token->line, operator_token->column
        );
    }
    
    return expr;
}

std::unique_ptr<ASTNode> Parser::logicalAnd() {
    auto expr = equality();
    
    while (match({TokenType::AND})) {
        Token* operator_token = tokens[current - 1].get();
        auto right = equality();
        expr = std::make_unique<BinaryOpNode>(
            std::move(expr), std::move(right), operator_token->type,
            operator_token->line, operator_token->column
        );
    }
    
    return expr;
}

std::unique_ptr<ASTNode> Parser::equality() {
    auto expr = comparison();
    
    while (match({TokenType::NOT_EQUAL, TokenType::EQUAL})) {
        Token* operator_token = tokens[current - 1].get();
        auto right = comparison();
        expr = std::make_unique<BinaryOpNode>(
            std::move(expr), std::move(right), operator_token->type,
            operator_token->line, operator_token->column
        );
    }
    
    return expr;
}

std::unique_ptr<ASTNode> Parser::comparison() {
    auto expr = term();
    
    while (match({TokenType::GREATER_THAN, TokenType::GREATER_EQUAL,
                  TokenType::LESS_THAN, TokenType::LESS_EQUAL})) {
        Token* operator_token = tokens[current - 1].get();
        auto right = term();
        expr = std::make_unique<BinaryOpNode>(
            std::move(expr), std::move(right), operator_token->type,
            operator_token->line, operator_token->column
        );
    }
    
    return expr;
}

std::unique_ptr<ASTNode> Parser::term() {
    auto expr = factor();
    
    while (match({TokenType::MINUS, TokenType::PLUS})) {
        Token* operator_token = tokens[current - 1].get();
        auto right = factor();
        expr = std::make_unique<BinaryOpNode>(
            std::move(expr), std::move(right), operator_token->type,
            operator_token->line, operator_token->column
        );
    }
    
    return expr;
}

std::unique_ptr<ASTNode> Parser::factor() {
    auto expr = unary();
    
    while (match({TokenType::DIVIDE, TokenType::MULTIPLY, TokenType::MODULO})) {
        Token* operator_token = tokens[current - 1].get();
        auto right = unary();
        expr = std::make_unique<BinaryOpNode>(
            std::move(expr), std::move(right), operator_token->type,
            operator_token->line, operator_token->column
        );
    }
    
    return expr;
}

std::unique_ptr<ASTNode> Parser::unary() {
    if (match({TokenType::NOT, TokenType::MINUS})) {
        Token* operator_token = tokens[current - 1].get();
        auto right = unary();
        return std::make_unique<UnaryOpNode>(
            std::move(right), operator_token->type,
            operator_token->line, operator_token->column
        );
    }
    
    return call();
}

std::unique_ptr<ASTNode> Parser::call() {
    auto expr = primary();
    
    while (true) {
        if (match({TokenType::LEFT_PAREN})) {
            expr = finishCall(std::move(expr));
        } else {
            break;
        }
    }
    
    return expr;
}

std::unique_ptr<ASTNode> Parser::finishCall(std::unique_ptr<ASTNode> callee) {
    auto call_node = std::make_unique<FunctionCallNode>(std::move(callee));
    
    if (!check(TokenType::RIGHT_PAREN)) {
        do {
            call_node->arguments.push_back(expression());
        } while (match({TokenType::COMMA}));
    }
    
    consume(TokenType::RIGHT_PAREN, "Expected ')' after arguments");
    return std::move(call_node);
}

std::unique_ptr<ASTNode> Parser::primary() {
    if (match({TokenType::NUMBER})) {
        Token* token = tokens[current - 1].get();
        return std::make_unique<LiteralNode>(token->lexeme, TokenType::NUMBER, 
                                           token->line, token->column);
    }
    
    if (match({TokenType::STRING})) {
        Token* token = tokens[current - 1].get();
        return std::make_unique<LiteralNode>(token->lexeme, TokenType::STRING,
                                           token->line, token->column);
    }
    
    if (match({TokenType::IDENTIFIER})) {
        Token* token = tokens[current - 1].get();
        return std::make_unique<IdentifierNode>(token->lexeme, token->line, token->column);
    }
    
    if (match({TokenType::LEFT_PAREN})) {
        auto expr = expression();
        consume(TokenType::RIGHT_PAREN, "Expected ')' after expression");
        return expr;
    }
    
    error("Expected expression");
    throw std::runtime_error("Expected expression");
}

std::string Parser::parseType() {
    bool is_optional = match({TokenType::QUESTION});
    
    if (match({TokenType::I32})) return is_optional ? "?i32" : "i32";
    if (match({TokenType::F64})) return is_optional ? "?f64" : "f64";
    if (match({TokenType::STRING_TYPE})) return is_optional ? "?string" : "string";
    if (match({TokenType::BOOL})) return is_optional ? "?bool" : "bool";
    if (match({TokenType::VOID})) return "void";
    
    if (match({TokenType::IDENTIFIER})) {
        std::string type_name = tokens[current - 1]->lexeme;
        
        // Check for array syntax
        if (match({TokenType::LEFT_BRACKET})) {
            consume(TokenType::RIGHT_BRACKET, "Expected ']' after '['");
            type_name += "[]";
        }
        
        return is_optional ? "?" + type_name : type_name;
    }
    
    error("Expected type");
    throw std::runtime_error("Expected type");
}

std::vector<Parameter> Parser::parseParameters() {
    std::vector<Parameter> parameters;
    
    if (!check(TokenType::RIGHT_PAREN)) {
        do {
            Token* name = consume(TokenType::IDENTIFIER, "Expected parameter name");
            consume(TokenType::COLON, "Expected ':' after parameter name");
            std::string type = parseType();
            
            parameters.emplace_back(name->lexeme, type, type[0] == '?');
        } while (match({TokenType::COMMA}));
    }
    
    return parameters;
}

} // namespace zen